Sources/BraveUI/SwiftUI/WebImageReader.swift

```diff
@@ -5,82 +5,70 @@
-        guard let self = self else { return }
-        self.image = image
-        self.error = error
-        if finished {
-          self.isFinished = true
-        }
-      })
+extension EnvironmentValues {
+  public var webImageDownloader: WebImageDownloaderType {
+    get { self[WebImageDownloaderKey.self] }
+    set { self[WebImageDownloaderKey.self] = newValue }
   }
+}
 
-  func cancel() {
-    operation?.cancel()
-    operation = nil
+extension SDWebImageManager: WebImageDownloaderType {
+  @MainActor public func downloadImage(url: URL) async -> (UIImage?, Error?) {
+    var operation: SDWebImageCombinedOperation?
+    return await withTaskCancellationHandler {
+      await withCheckedContinuation { continuation in
+        operation = loadImage(with: url, progress: nil) { image, _, error, _, _, _ in
+          continuation.resume(returning: (image, error))
+        }
+      }
+    } onCancel: { [operation] in
+      operation?.cancel()
+    }
   }
   
-  func load(base64Str: String, options: [SDImageCoderOption: Any] = [:]) {
-    guard base64Str.hasPrefix("data:image/") else { return }
-    guard let dataString = base64Str.separatedBy(",").last else { return }
-    
-    let data = Data(base64Encoded: dataString, options: .ignoreUnknownCharacters)
-    for coder in supportedCoders where coder.canDecode(from: data) {
-      image = coder.decodedImage(with: data, options: options)
-      break
-    }
+  public func imageFromData(data: Data) async -> UIImage? {
+    SDImageCodersManager.shared.decodedImage(with: data)
   }
 }
 
 public struct WebImageReader<Content: View>: View {
-  @StateObject private var imageManager: WalletWebImageManager = .init()
-  var url: URL?
-  var options: SDWebImageOptions
-  var coderOptions: [SDImageCoderOption: Any]
+  var url: URL
+  
+  @Environment(\.webImageDownloader) private var imageDownloader: WebImageDownloaderType
+  
+  @State private var image: UIImage?
+  @State private var error: Error?
 
-  private var content: (_ image: UIImage?, _ isFinished: Bool) -> Content
+  private var content: (_ image: UIImage?, _ error: Error?) -> Content
 
   public init(
-    url: URL?,
-    options: SDWebImageOptions = [],
-    coderOptions: [SDImageCoderOption: Any] = [:],
-    @ViewBuilder content: @escaping (_ image: UIImage?, _ isFinished: Bool) -> Content
+    url: URL,
+    @ViewBuilder content: @escaping (_ image: UIImage?, _ error: Error?) -> Content
   ) {
     self.content = content
     self.url = url
-    self.options = options
-    self.coderOptions = coderOptions
   }
 
   public var body: some View {
-    content(imageManager.image, imageManager.isFinished)
-      .onAppear {
-        if let urlString = url?.absoluteString {
-          if urlString.hasPrefix("data:image/") {
-            imageManager.load(base64Str: urlString)
-          } else if !imageManager.isFinished {
-              imageManager.load(url: url, options: options)
-          }
+    content(image, error)
+      .id(url)
```

Comment:
This may need to be _after_ the `task`? Not sure if `id` will apply correctly to the modifiers after it

Reply:
`id` is now after `task` in [8e4439d](https://github.com/brave/brave-ios/pull/7881/commits/8e4439db703e9ca9718fade5121652e75ebc5e68)